app/src/animation/animation_ripple.c

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_animation_ripple
+
+#include <stdlib.h>
+#include <math.h>
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+#include <drivers/animation.h>
+
+#include <zmk/animation.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+struct animation_ripple_event {
+    size_t pixel_id;
+    uint16_t distance;
+    uint8_t counter;
+};
+
+struct animation_ripple_config {
+    struct zmk_color_hsl *color_hsl;
+    size_t *pixel_map;
+    size_t pixel_map_size;
+    size_t event_buffer_size;
+    uint8_t blending_mode;
+    uint8_t distance_per_frame;
+    uint8_t ripple_width;
+    uint8_t event_frames;
+};
+
+struct animation_ripple_data {
+    struct zmk_color_rgb color_rgb;
+    struct animation_ripple_event *event_buffer;
+    size_t events_start;
+    size_t events_end;
+    size_t num_events;
+    bool is_active;
+};
+
+static int animation_ripple_on_key_press(const struct device *dev, const zmk_event_t *event) {
+    const struct animation_ripple_config *config = dev->config;
+    struct animation_ripple_data *data = dev->data;
+
+    const struct zmk_position_state_changed *pos_event;
+
+    if (!data->is_active) {
+        return 0;
+    }
+
+    if ((pos_event = as_zmk_position_state_changed(event)) == NULL) {
+        // Event not supported.
+        return -ENOTSUP;
+    }
+
+    if (!pos_event->state) {
+        // Don't track key releases.
+        return 0;
+    }
+
+    if (data->num_events == config->event_buffer_size) {
+        // Event buffer is full - new key press events are dropped.
+        return -ENOMEM;
+    }
+
+    data->event_buffer[data->events_end].pixel_id =
+        zmk_animation_get_pixel_by_key_position(pos_event->position);
+    data->event_buffer[data->events_end].distance = 0;
+    data->event_buffer[data->events_end].counter = 0;
+
+    data->events_end = (data->events_end + 1) % config->event_buffer_size;
+    data->num_events += 1;
+
+    zmk_animation_request_frames(config->event_frames);
+
+    return 0;
+}
+
+static void animation_ripple_render_frame(const struct device *dev, struct animation_pixel *pixels,
+                                          size_t num_pixels) {
+    const struct animation_ripple_config *config = dev->config;
+    struct animation_ripple_data *data = dev->data;
+
+    size_t *pixel_map = config->pixel_map;
+
+    size_t i = data->events_start;
+
+    while (i != data->events_end) {
+        struct animation_ripple_event *event = &data->event_buffer[i];
+
+        // Render all pixels for each event
+        for (int j = 0; j < config->pixel_map_size; ++j) {
+            uint8_t pixel_distance = zmk_animation_get_pixel_distance(event->pixel_id, j);
```

Comment:
Shouldn't this be `zmk_animation_get_pixel_distance(event->pixel_id, pixel_map[j]);`? Since we are using the pixel map here?

Reply:
Nice spot! You're right... Miraculously these were somehow identical on my test board with the current config, lol. Thanks!